src/transformers/models/jamba/modeling_jamba.py

```diff
@@ -0,0 +1,2071 @@
+        output_attentions (`bool`, *optional*):
+            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+            tensors for more detail.
+        output_hidden_states (`bool`, *optional*):
+            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
+            more detail.
+        output_router_logits (`bool`, *optional*):
+            Whether or not to return the logits of all the routers. They are useful for computing the router loss, and
+            should not be returned during inference.
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
+
+
+@add_start_docstrings(
+    "The bare Jamba Model outputting raw hidden-states without any specific head on top.",
+    JAMBA_START_DOCSTRING,
+)
+# Adapted from transformers.models.mistral.modeling_mistral.MistralModel with MISTRAL->JAMBA, Mistral->Jamba
+class JambaModel(JambaPreTrainedModel):
+    """
+    Transformer decoder consisting of *config.num_hidden_layers* layers. Each layer is a [`JambaDecoderLayer`]
+
+    Args:
+        config: JambaConfig
+    """
+
+    def __init__(self, config: JambaConfig):
+        super().__init__(config)
+        self.padding_idx = config.pad_token_id
+        self.vocab_size = config.vocab_size
+
+        # The maximum sequence length that this model is intended to be used with. Doesn't have any real effect
+        self.n_ctx = config.n_ctx
+
+        self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size, self.padding_idx)
+
+        # init each model layer, decide if it's mamba/attention and has experts or not
+        decoder_layers = []
+        for i in range(config.num_hidden_layers):
+            is_attn = True if (i - self.config.attn_layer_offset) % self.config.attn_layer_period == 0 else False
+            is_expert = True if (i - self.config.expert_layer_offset) % self.config.expert_layer_period == 0 else False
+
+            num_experts = self.config.num_experts if is_expert else 1
+            layer_class = JambaAttentionDecoderLayer if is_attn else JambaMambaDecoderLayer
+            decoder_layers.append(layer_class(config, num_experts=num_experts, layer_idx=i))
+
+        if not any(isinstance(layer, JambaAttentionDecoderLayer) for layer in decoder_layers):
+            raise ValueError("At least one layer in the decoder must be an attention layer")
+        self._attn_layer_index = [isinstance(layer, JambaAttentionDecoderLayer) for layer in decoder_layers].index(
+            True
+        )
+
+        if not any(isinstance(layer, JambaMambaDecoderLayer) for layer in decoder_layers):
+            raise ValueError("At least one layer in the decoder must be a Mamba layer")
+        self._mamba_layer_index = [isinstance(layer, JambaMambaDecoderLayer) for layer in decoder_layers].index(True)
```

Comment:
since we have the info here, initializing the caches in the forward makes sense (the hybrid cache), to then pass it (like mamba cache is initialized).